### PR TITLE
formula_auditor: Having `HOMEBREW_PREFIX` in `keg_only` reasons is bad

### LINKED ISSUE #14996: Handle `HOMEBREW_PREFIX` references in `keg_only` reasons
### Verification

- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### Provide a detailed description of the proposed feature

`brew` should handle `HOMEBREW_PREFIX` references in `keg_only` reasons. Examples of ways to handle this are:
1. Relocate the `HOMEBREW_PREFIX` reference
2. Disallow `HOMEBREW_PREFIX` references by an audit

The above list may not be exhaustive.

### What is the motivation for the feature?

This can lead to confusing `caveats` when installing from the API. For example, the output of `brew info socket_vmnet` produces:

```
==> Caveats
To install an optional launchd service, run the following command (sudo is necessary):
sudo brew services start socket_vmnet

socket_vmnet is keg-only, which means it was not symlinked into /opt/homebrew,
because /usr/local/bin is often writable by a non-admin user.
```

### How will the feature be relevant to at least 90% of Homebrew users?

Fewer confusing caveats.

### What alternatives to the feature have been considered?

Doing nothing.

### PR DESCRIPTION
**Pull request by issyl0**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

- Fixes #14996.
- The API JSON is generated with a `/usr/local` Homebrew prefix, but frequently now users have `/opt/homebrew` as their prefix. Since formulae `keg_only` reasons are generated by the API, this can lead to the confusing messaging that follows:

```
socket_vmnet is keg-only, which means it was not symlinked into /opt/homebrew,
because /usr/local/bin is often writable by a non-admin user.
```

- With this change, that formula as it is now will fail `brew audit`, prompting to remove the prefix reference.

```
❯ HOMEBREW_NO_INSTALL_FROM_API=1 brew audit --strict socket_vmnet
socket_vmnet:
  * `keg_only` reason should not include `HOMEBREW_PREFIX` as it provides confusing output.
Error: 1 problem in 1 formula detected
```

### CODE REVIEW

File: Library/Homebrew/formula_auditor.rb [hunk 1]
```diff
@@ -466,6 +466,16 @@ def audit_elasticsearch_kibana
...
     end
 
+    def audit_keg_only_reason
+      return unless @core_tap
+      return unless formula.keg_only?
+
+      keg_only_message = text.to_s.match(/keg_only\s+["'](.*)["']/).captures&.first
+      return unless keg_only_message&.include?("HOMEBREW_PREFIX")
+
+      problem "`keg_only` reason should not include `HOMEBREW_PREFIX` as it provides confusing output."
```
Comment (issyl0) [inline Library/Homebrew/formula_auditor.rb:476]: The word "provides" doesn't seem quite right here. "Shows"? "Gives"?

Comment (carlocab) [reply Library/Homebrew/formula_auditor.rb:476]:
```suggestion
      problem "`keg_only` reason should not include `HOMEBREW_PREFIX` as it creates confusing `brew info` output."
```
?

**Review by carlocab — COMMENTED** [review]
I'm actually also not sure on this one! (Sorry, I know I must be giving mixed messages.)

On the one hand, an audit would be nice to make sure we never break `keg_only` reasons. I suspect it's bound to happen again at some point. On the other, you're right that it doesn't really happen all that often so an audit may be overkill.

I'll let someone with a more neutral perspective make a call here!

**Comment by carlocab** [comment]
Oh, also: it may be nicer to just disallow all interpolation in `keg_only` reasons (if we want an audit at all!). Some interpolation may work, but it's easier to just disallow all of them to be sure.

**Comment by issyl0** [comment]
> I'm actually also not sure on this one! (Sorry, I know I must be giving mixed messages.)

You are indeed giving mixed messages. 😂

**Comment by Rylan12** [comment]
We solved this problem with cask caveats, right? Can we use a similar method? I know we used to use the `to_h_gsubs` methods to just replace `HOMEBREW_PREFIX` with a placeholder (and then re-replace when loading from the API). I think we replaced that with a new method that also supported `appdir` but I'll need to look at the code to remember

**Comment by issyl0** [comment]
@Rylan12 Thanks! I had no idea that existed. I'll look into that for `keg_only` reasons.

**Comment by issyl0** [comment]
Going to do https://github.com/Homebrew/brew/pull/15003#issuecomment-1474882176 and not make this an audit. Thanks for the pointer, Rylan!

**Comment by issyl0** [comment]
Re-opening since this might be a better, simpler solution after all than #15004!

**Review by MikeMcQuaid — APPROVED** [review]

File: Library/Homebrew/formula_auditor.rb [hunk 2]
```diff
@@ -466,6 +466,16 @@ def audit_elasticsearch_kibana
               "They must not be upgraded to version 7.11 or newer."
     end
 
+    def audit_keg_only_reason
+      return unless @core_tap
+      return unless formula.keg_only?
+
+      keg_only_message = text.to_s.match(/keg_only\s+["'](.*)["']/).captures&.first
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_auditor.rb:473]:
```suggestion
      keg_only_message = text.to_s.match(/keg_only\s+["'](.*)["']/)&.captures&.first
```

or maybe

```suggestion
      keg_only_message = text.to_s.match(/keg_only\s+["'](.*)["']/)&.captures.first
```

**Review by MikeMcQuaid — COMMENTED** [review]
Whoops, wee CI failure here.

File: Library/Homebrew/formula_auditor.rb [hunk 3]
```diff
@@ -466,6 +466,16 @@ def audit_elasticsearch_kibana
               "They must not be upgraded to version 7.11 or newer."
     end
 
+    def audit_keg_only_reason
+      return unless @core_tap
+      return unless formula.keg_only?
+
+      keg_only_message = text.to_s.match(/keg_only\s+["'](.*)["']/)&.captures.first
```
Comment (issyl0) [inline Library/Homebrew/formula_auditor.rb:473]:
```suggestion
      keg_only_message = text.to_s.match(/keg_only\s+["'](.*)["']/)&.captures&.first
```

**Pull request merged**